text/0004-type-deduction.md

```diff
@@ -0,0 +1,214 @@
+
+### Special case 2: Binary which depends on the result type of another expression
+```shell Example
+ls | where size < $a * ($b * $c); kill $b $c
+$a -> SyntaxShape::Unit
+$b -> SyntaxShape::Int
+$c -> SyntaxShape::Int
+```
+Same as above. But this time $a also depends on the result type of ($b * $c). This dependency has to be postponed, as $b nor $c can be deduced at the point in time of traversing the AST.
+
+
+As soon as the AST is traversed, the dependencies have to be resolved.
+Please note: One might think of a dependency as a edge in a directed graph, with variables as nodes. Every Node with no outgoing edge is completly deduced. Every Node with an edge towards such a completly deduced node can then be inferred. Nodes with cycles (e.G. `ls | where size < $a * $b` $a depends on $b, $b depends on $a) can't be deduced completly.
+It is in question what to return here best dependend on the operator in use. One might solves this elegantly by handling
+this case the same as for optional positional arguments with dependencies between variables.
+(Note: One might also think about this not as a graph problem, but as an CSP.)
+
+## Merging of deductions.
+Any variable can occur multiple times within the block. In each position the variable can have different possible Types.
+```shell Example
+config | where size < 1kb * $a; kill $a
+$a -> SyntaxShape::Int
+```
+In the first position, $a might be SyntaxShape::Int or SyntaxShape::Number. In the second it is of SyntaxShape::Int.
+The result set of possible types for a variable can be computed as the set intersection between the already deduced types and the new deduced types. If the intersection is empty, it means that the variable has usages with different and non compatible types. An error is returned.
+
+```rust
+fn checked_insert(existing_deductions, new_deductions) -> Result<set_of_deductions, ShellError>{
+    if existing_deductions == None return new_deductions
+    else return set_intersection(existing_deductions, new_deductions)?
+}
+
+existing_deductions = checked_insert(existing_deductions, new_deductions);
+```
+
+Special cases arise if one (or both) sets contains SyntaxShape::Any. SyntaxShape::Any is a placeholder for any possible type. The above pseudo code has to be changed to: 
+```rust
+fn checked_insert(existing_deductions, new_deductions) -> Result<set_of_deductions, ShellError>{
+    if existing_deductions == None return new_deductions
+    else match (has_any(existing_deductions), has_any(new_deductions)){
+        (true, true) -> set_union_including_any(existing_deductions, new_deductions)
+        (true, false) -> set_union_excluding_any(new_deductions, set_intersection(existing_deductions, new_deductions))
+        (false, true) -> set_union_excluding_any(existing_deductions, set_intersection(existing_deductions, new_deductions))
+        (false, false) -> set_intersection(existing_deductions, new_deductions)?
+    }
+}
```

Comment:
I wonder if this would be simpler to reason about if we expanded `SyntaxShape::Any` into the set of types it can represent, and then did `checked_insert` like above (just a set intersection) 🤔

Reply:
Good hint :)

Reply:
I was just wondering that too. A kind of `SyntaxShape::OneOf(a, b, c)` form.

Reply:
> I wonder if this would be simpler to reason about if we expanded `SyntaxShape::Any` into the set of types it can represent, and then did `checked_insert` like above (just a set intersection) thinking

After thinking this through, I am not sure whether it would simplify the implementation.
Looking closes at the pseudo code, it is obvious that:
```rust
        (true, false) -> set_union_excluding_any(new_deductions, set_intersection(existing_deductions, new_deductions))
```
is the same as:
```rust
        (true, false) -> new_deductions
```

The reason I have written it down like this is, that meta information (where was this deduction be made?) needs to get merged. If `SyntaxShape::Any` would be expanded into every type, handling the meta data correctly might be a pain in the concrete implementation. 


> I was just wondering that too. A kind of `SyntaxShape::OneOf(a, b, c)` form.
OneOf should then probably contain a vector of SyntaxShape variants? From my point of view this is almost the same as a vector of SyntaxShape's. It may make the code a little more declarative, but at the added cost of having a different size for the SyntaxShape enum. I would like to stick with the current pseudo code.